taichi/python/export_lang.cpp

```diff
@@ -622,55 +624,96 @@ void export_lang(py::module &m) {
+          } else if (tag == aot::ArgKind::kScalar) {
+            auto expected_dtype = arg.dtype();
+            insert_scalar_arg(arg_name, expected_dtype, pyarg);
+          } else if (tag == aot::ArgKind::kMatrix) {
+            auto type_id = arg.dtype()->as<PrimitiveType>()->type;
+            switch (type_id) {
+              case PrimitiveTypeID::f16: {
+                auto arr = pyarg.cast<py::array_t<float32>>();
+                py::buffer_info buffer_info = arr.request();
+                auto length = buffer_info.size;
+                auto ptr = reinterpret_cast<intptr_t>(buffer_info.ptr);
+
+                std::unique_ptr<char[]> data(new char[128]);
+                for (uint32_t i = 0; i < length; i++) {
+                  uint16 half = fp16_ieee_from_fp32_value(
+                      reinterpret_cast<float32 *>(ptr)[i]);
+                  reinterpret_cast<uint16 *>(data.get())[i] = half;
+                }
+                matrix_buffers.emplace_back(std::move(data));
+
+                matrices.emplace_back(Matrix(
+                    length, arg.dtype(),
+                    reinterpret_cast<intptr_t>(matrix_buffers.back().get())));
+                args.insert({arg_name, aot::IValue::create(matrices.back())});
+                break;
+              }
+#define PER_C_TYPE(type, ctype)                                           \
+  case PrimitiveTypeID::type: {                                           \
+    auto arr = pyarg.cast<py::array_t<ctype>>();                          \
+    py::buffer_info buffer_info = arr.request();                          \
+    auto length = buffer_info.size;                                       \
+    auto ptr = reinterpret_cast<intptr_t>(buffer_info.ptr);               \
+                                                                          \
+    std::unique_ptr<char[]> data(new char[128]);                          \
+    std::memcpy(data.get(), reinterpret_cast<char *>(ptr),                \
+                sizeof(ctype) * length);                                  \
+    matrix_buffers.emplace_back(std::move(data));                         \
+                                                                          \
+    matrices.emplace_back(                                                \
+        Matrix(length, arg.dtype(),                                       \
+               reinterpret_cast<intptr_t>(matrix_buffers.back().get()))); \
+    args.insert({arg_name, aot::IValue::create(matrices.back())});        \
+    break;                                                                \
+  }
+#include "taichi/inc/data_type_with_c_type.inc.h"
+#undef PER_C_TYPE
+              default:
+                TI_ERROR("Unsupported scalar type {}", type_id);
+            }
+          } else {
+            TI_NOT_IMPLEMENTED;
+          }
+        }
+        self->jit_run(compile_config, args);
+        matrix_buffers.clear();
```

Comment:
nit: do you need to clear these manually?

Reply:
Solved.